### PR TITLE
chore(quil-rs): benchmarking expression simplification

### LINKED ISSUE #269: Expression simplification needs benchmarking
In order to have data for or against things like #267, we need some benchmarking of expression simplification.

### PR DESCRIPTION
**Pull request by genos**
Closes #269.

<details><summary>I created the <code>expressions.txt</code> file with additions to the <code>Expression</code> test suite (click to see more)</summary>

```diff
diff --git a/quil-rs/Cargo.toml b/quil-rs/Cargo.toml
index 4e998ca..deebdc4 100644
--- a/quil-rs/Cargo.toml
+++ b/quil-rs/Cargo.toml
@@ -22,6 +22,8 @@ nom_locate = "4.0.0"
 num-complex = "0.4.0"
 once_cell = "1.17.1"
 petgraph = "0.6.2"
+rand = "0.8.5"
+rand_chacha = "0.3.1"
 regex = "1.7.1"
 serde = { version = "1.0.125", features = ["derive"] }
 strum.workspace = true
diff --git a/quil-rs/src/expression/mod.rs b/quil-rs/src/expression/mod.rs
index 1bb8a17..cb266f1 100644
--- a/quil-rs/src/expression/mod.rs
+++ b/quil-rs/src/expression/mod.rs
@@ -1121,4 +1121,128 @@ mod tests {
             assert_eq!(format_complex(x), *s);
         }
     }
+
+    use rand::{distributions::Alphanumeric, Rng, SeedableRng};
+    use rand_chacha::ChaCha20Rng;
+    use std::{
+        fs::File,
+        io::{BufWriter, Write},
+    };
+
+    fn build(rng: &mut impl Rng, depth: u64) -> Expression {
+        if depth == 0 {
+            match rng.gen_range(0..4) {
+                0 => addr(rng),
+                1 => number(rng),
+                2 => Expression::PiConstant,
+                3 => var(rng),
+                _ => unreachable!(),
+            }
+        } else {
+            let d = depth - 1;
+            match rng.gen_range(0..3) {
+                0 => func(rng, d),
+                1 => infix(rng, d),
+                2 => prefix(rng, d),
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    fn _name(rng: &mut impl Rng) -> String {
+        let len = rng.gen_range(1..=10);
+        let mut name = rng
+            .sample_iter(&Alphanumeric)
+            .take(len)
+            .map(char::from)
+            .collect::<String>();
+        while ReservedToken::from_str(&name).is_ok()
+            || name.to_lowercase() == "nan"
+            || name.chars().next().map(|c| c.is_digit(10)).unwrap_or(false)
+        {
+            name = rng
+                .sample_iter(&Alphanumeric)
+                .take(len)
+                .map(char::from)
+                .collect();
+        }
+        name
+    }
+
+    fn addr(rng: &mut impl Rng) -> Expression {
+        Expression::Address(MemoryReference {
+            name: _name(rng),
+            index: rng.gen(),
+        })
+    }
+
+    fn number(rng: &mut impl Rng) -> Expression {
+        Expression::Number(num_complex::Complex::<f64> {
+            re: rng.gen(),
+            im: rng.gen(),
+        })
+    }
+
+    fn var(rng: &mut impl Rng) -> Expression {
+        Expression::Variable(_name(rng))
+    }
+
+    fn func(rng: &mut impl Rng, depth: u64) -> Expression {
+        let function = match rng.gen_range(0..5) {
+            0 => ExpressionFunction::Cis,
+            1 => ExpressionFunction::Cosine,
+            2 => ExpressionFunction::Exponent,
+            3 => ExpressionFunction::Sine,
+            4 => ExpressionFunction::SquareRoot,
+            _ => unreachable!(),
+        };
+        Expression::FunctionCall(FunctionCallExpression {
+            function,
+            expression: build(rng, depth).into(),
+        })
+    }
+
+    fn infix(rng: &mut impl Rng, depth: u64) -> Expression {
+        let operator = match rng.gen_range(0..5) {
+            0 => InfixOperator::Caret,
+            1 => InfixOperator::Plus,
+            2 => InfixOperator::Minus,
+            3 => InfixOperator::Slash,
+            4 => InfixOperator::Star,
+            _ => unreachable!(),
+        };
+        Expression::Infix(InfixExpression {
+            left: build(rng, depth).into(),
+            operator,
+            right: build(rng, depth).into(),
+        })
+    }
+
+    fn prefix(rng: &mut impl Rng, depth: u64) -> Expression {
+        let operator = if rng.gen() {
+            PrefixOperator::Plus
+        } else {
+            PrefixOperator::Minus
+        };
+        Expression::Prefix(PrefixExpression {
+            operator,
+            expression: build(rng, depth).into(),
+        })
+    }
+
+    #[test]
+    fn write_expressions_for_simplifying() {
+        // Doing this by hand rather than with proptest because
+        //   - more reproducbility
+        //   - I want big expressions
+        let mut rng = ChaCha20Rng::seed_from_u64(27586845); // random.org
+        let file = File::create("expressions.txt").expect("should be able to make a file");
+        let mut buf = BufWriter::new(file);
+        for _ in 0..25 {
+            let e = build(&mut rng, 5);
+            let p = parenthesized(&e);
+            assert!(Expression::from_str(&p).is_ok());
+            assert!(writeln!(buf, "{}", &p).is_ok());
+        }
+    }
 }
```

</details>

I'm not including them here because they're superfluous.

### CODE REVIEW

File: quil-rs/benches/simplification.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+fn simplify(e: Expression) -> Expression {
+    e.into_simplified()
+}
+
+fn benchmark_expression_simplification(c: &mut Criterion) {
+    EXPRESSIONS.iter().enumerate().for_each(|(i, e)| {
+        c.bench_function(&format!("expression_{i}"), |b| {
+            b.iter(|| black_box(simplify(e.clone())))
+        });
+    })
```
Comment (kalzoo) [inline quil-rs/benches/simplification.rs:22]: if we rely on the vec index to key the inputs, that means we can't remove, change, or reorder them without tampering with our results. maybe a prefix on each line with the index, then lines can be removed without consequence?

**Comment by kalzoo** [comment]
> I'm not including them here because they're superfluous.

I think you should! That's useful and we may want to invoke it again in the future, shame to throw that away. Rather than a test, maybe `src/bin/generate_test_expressions.rs` with a docstring to explain its purpose?

**Review by kalzoo — CHANGES_REQUESTED** [review]
Thanks @genos ! Please also add this benchmark suite to the github actions CI (`--bench` flag) so they get captured and tracked over time.

**Comment by genos** [comment]
Thanks @kalzoo! In 9496778, I put the expression generation binary in `quil-rs/examples` instead of `bin`, since examples use development dependencies like tests; this meant that I could add `clap` and `rand` to `dev-dependencies` without requiring them for users of the library.

**Review by kalzoo — APPROVED** [review]
This looks great! I'm not concerned about the multiple-versions issue, that's cargo working as designed. Maybe add `multiple-versions = "warn"` to the deny config to flip the pipeline green?

**Comment by genos** [comment]
Turns out that updating `criterion` in f7fb9c2 fixed `cargo-deny`'s complaint. I just merged `main` into here, too, so I think this is good to go? cc @kalzoo

**Review by kalzoo — APPROVED** [review]

**Pull request merged**